src/garage/sampler/config.py

```diff
@@ -0,0 +1,217 @@
+    `env_update` into `obtain_samples`.
+
+    Args:
+        config(SamplerConfig): The config this function is stored in. This
+            argument is ignored, but present to match the API set by
+            `SamplerConfig`.
+        worker_number(int): The number of the worker this update is occurring
+            in. This argument is ignored, but present to match the API set by
+            `SamplerConfig`.
+        env(gym.Env): The environment to update.
+        env_update(gym.Env or None): The environment to replace it with.
+
+    Returns:
+        The updated environment.
+
+    """
+    if env_update is not None:
+        return env_update
+    else:
+        return env
+
+
+def default_agent_update_fn(config, worker_number, agent, agent_update):
+    """Update an agent, assuming it implements garage.tf.policies.Policy2.
+
+    Args:
+        config(SamplerConfig): The config this function is stored in. This
+            argument is ignored, but present to match the API set by
+            `SamplerConfig`.
+        worker_number(int): The number of the worker this update is occurring
+            in. This argument is ignored, but present to match the API set by
+            `SamplerConfig`.
+        agent(garage.tf.policies.Policy2): The agent to update.
+        agent_update(dict[str, np.array]): Parameters to agent, which should
+            have been generated by calling `policy.get_param_values`.
+
+    Returns:
+        The updated agent.
+
+    """
+    agent.set_param_values(agent_update)
+    return agent
+
+
+def default_worker_init_fn(config, worker_number):
+    """Initialize a worker.
+
+    Args:
+        worker_number(int): The number of the worker this update is occurring
+            in. This argument is used to set a different seed for each worker.
+        config(SamplerConfig): The config this function is stored in. Used to
+            get the main seed, on which all other seeds used by the sampler are
+            based.
+
+    """
+    deterministic.set_seed(config.seed + worker_number)
+
+
+def identity_function(x):
+    """Do nothing.
+
+    This function can exists so it can be pickled.
+    """
+    return x
+
+
+class SamplerConfig:
+    """An object for configuring the sampler.
+
+    The intent is that this object should be sufficient to avoid subclassing
+    the sampler. Instead of subclassing the sampler for e.g. a specific
+    backend, implement a specialized SamplerConfig (or specify appropriate
+    functions to this one). Not that this object must be picklable, since it
+    may be passed to workers. However, its fields individually need not be.
+
+    All arguments to this type must be passed by keyword.
+
+    Args:
+        seed(int): The seed to use to intialize random number generators.
+        n_workers(int): The number of workers to use.
+        max_path_length(int): The maximum length paths which will be sampled.
+        worker_init_fn((int, SamplerConfig) -> None): Function to run in
```

Comment:
worker number is an odd thing to be carrying around/passing everywhere, but I understand the intent. passing `SamplerConfig` to `worker_init_fn` and `rollout`, sometimes in addition to `worker_number` seems an odd asymmetry to me.

more importantly, who gets a `SamplerConfig` and who gets a `worker_number` (or both) i find a little hard to predict from just reading the API.

what if you split the roles here into `SamplerConfig` which is global/initialization and `SamplingContext` which is generated by `Sampler` and created/managed per-worker. for instance, the context could include the random seed this worker should use (rather than expecting the worker to derive it from the global seed), the worker number, perhaps which resources this worker is allowed to use, etc.

Reply:
I ended up doing something like this (`WorkerFactory` and `Worker`). I do think it makes sense to rename these though, since they're not exactly workers in the sense most people probably expect.